### PR TITLE
[masterbots.ai] feat: Create Password Recovery

### PR DESCRIPTION
**Pull request by Bran18**
## Summary

Introduce a password recovery feature with endpoints for requesting and resetting passwords, along with UI components for user interaction. Implement email functionality to facilitate password reset requests. Update Hasura metadata to support the new feature and refactor Taskfile.yml for improved consistency.

New Features:
- Implement password recovery feature, including endpoints for requesting and resetting passwords.
- Add user interface components for 'Forgot Password' and 'Reset Password' forms.
- Introduce email functionality to send password reset emails using nodemailer.

Enhancements:
- Update Hasura metadata to include new tables and relationships for managing password reset tokens.
- Refactor Taskfile.yml to use 'docker compose' instead of 'docker-compose' for consistency.

**Note: MAILTRAP_USER and MAILTRAP_PASS are added as env variables and are crucial for testing**

## Release Notes

- **New Features**
  - Introduced password reset functionality with new API endpoints for requesting and processing password resets.
  - Added user interface components for "Forgot Password" and "Reset Password" processes.
  - Implemented a password strength meter to assist users in creating strong passwords.
  - Enhanced email sending capability for password reset notifications.

- **Bug Fixes**
  - Improved error handling and response messages for password reset operations.

- **Chores**
  - Added environment variable for Hasura GraphQL admin secret.

- **Dependencies**
  - Added `nodemailer` for email handling functionality.

### Preview video:

https://github.com/user-attachments/assets/449a5f83-0918-40b2-b346-f9ac530a34a5

### Password Strength - UX

https://github.com/user-attachments/assets/bdfb6220-ec5a-4da1-93ad-38945121e475

### Password Reveal - UX

https://github.com/user-attachments/assets/8a8cac00-6240-4f78-88f5-15f345ff6e55

### CODE REVIEW

File: apps/masterbots.ai/app/api/auth/reset-password/route.ts [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+      return NextResponse.json(
+        { error: 'Invalid or expired reset token' },
+        { status: 400 }
+      )
+    }
+
+    const userId = userToken[0].userId
+
+    // * Hash the new password
+    const salt = await bcryptjs.genSalt(10)
```
[the remark this answers is not in the document]
Comment (Bran18) [reply apps/masterbots.ai/app/api/auth/reset-password/route.ts:51]: True, but I'm keeping the bcryptjs.genSalt in 10 as the sign-in and register API route has it.

File: Taskfile.yml [hunk 2]
```diff
@@ -4,27 +4,22 @@ dotenv: [".env"]
...
     cmds:
-      - hasura console --project apps/hasura 
-
-  # [First Step When First Boot] Run the hasura backend (docker desktop must be started first)
-  boot-no-migrate:
-    cmds:
-      - docker-compose up -d --build database
-      - sleep 10
-      - docker-compose up -d --build hasura
+      - hasura console --project apps/hasura --admin-secret ${HASURA_GRAPHQL_ADMIN_SECRET}
```
[the remark this answers is not in the document]
Comment (Bran18) [reply Taskfile.yml:15]: In the newest version of docker, the flag wasn't working for all the commands

File: packages/mb-genql/package.json [hunk 3]
```diff
@@ -6,7 +6,7 @@
   "main": "./src/index.ts",
   "types": "./src/index.ts",
   "scripts": {
-    "gen": "genql --endpoint https://dev-api.masterbots.ai/v1/graphql --output ./generated -H 'content-type: application/json' -H 'x-hasura-admin-secret: 7916dce3ec9736725aa46ee1f99b8bb8'",
+    "gen": "genql --endpoint http://localhost:8080/v1/graphql --output ./generated -H 'content-type: application/json' -H 'x-hasura-admin-secret: lfg'",
```
[the remark this answers is not in the document]
Comment (Bran18) [reply packages/mb-genql/package.json:9]: @AndlerRL is this normal?

[the remark this answers is not in the document]
Comment (AndlerRL) [reply packages/mb-genql/package.json:9]: As for our scenario, it is. The old secret was previously securely updated and the new secret is a local secret that does not represent any harm, due it a made-up secret that doesn't exist on our remote services.

File: apps/masterbots.ai/app/auth/reset-password/page.tsx [hunk 4]
```diff
@@ -0,0 +1,17 @@
+import { Suspense } from 'react'
+import ResetPasswordForm from '@/components/auth/reset-password-form'
+
+export default function ResetPasswordPage({
+  searchParams
+}: {
+  searchParams: { token: string }
+}) {
```
[the remark this answers is not in the document]
Comment (Bran18) [reply apps/masterbots.ai/app/auth/reset-password/page.tsx:8]: What is the best way to apply sanitization to this token? @coderabitai

**Review by AndlerRL — APPROVED** [review]
Excellent work! :rocket: The additional functionalities to verify PW strength feature and the UX improvements looks very good and functionality looks good too :smile: I notice though some few improvements that we can do before to do the merge.  After those updates, we are good to go!

File: apps/masterbots.ai/app/api/auth/reset-password/route.ts
[the remark this answers is not in the document]
Comment (Bran18) [reply apps/masterbots.ai/app/api/auth/reset-password/route.ts:1]: I 100% agree on this one. I thought about creating a new profile in Hasura for this PW reset and the other related features, but your idea of Drizzle will be much better! I believe we can even protect the different APIs we currently have (auth-wordware for example) with the Drizzle implementation. 

Tks Andler

File: apps/masterbots.ai/components/auth/reset-password-form.tsx [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+import { Input } from '@/components/ui/input'
+import { Label } from '@/components/ui/label'
+import { toast } from 'react-hot-toast'
+import { useRouter } from 'next/navigation'
+import PasswordStrengthMeter from '@/components/shared/password-strength-meter'
+import { isPasswordStrong } from '@/lib/password'
+import { Eye, EyeOff } from 'lucide-react'
+
+export default function ResetPasswordForm({ token }: { token: string }) {
+  const [password, setPassword] = useState('')
```
[the remark this answers is not in the document]
Comment (Bran18) [reply apps/masterbots.ai/components/auth/reset-password-form.tsx:14]: Done we are now using `{ useSetState } from 'react-use'`

**Comment by Bran18** [comment]
> Excellent work! 🚀 The additional functionalities to verify PW strength feature and the UX improvements looks very good and functionality looks good too 😄 I notice though some few improvements that we can do before to do the merge. After those updates, we are good to go!

Thaks @AndlerRL all suggestions are done 🚀

**Comment by AndlerRL** [comment]
[Create Password Reset UI/UX + Verification](https://trello.com/c/O6lbse93/245-create-password-reset-ui-ux-verification)

**Pull request merged**